cmd/crdvalidator/internal/handlers/crd.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	"github.com/operator-framework/rukpak/cmd/crdvalidator/internal/crd"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// +kubebuilder:webhook:path=/validate-crd,mutating=false,failurePolicy=fail,groups="",resources=customresourcedefinitions,verbs=create;update,versions=v1,name=crd-validation-webhook.io
+
+// CrdValidator houses a client, decoder and Handle function for ensuring
+// that a CRD create/update request is safe
+type CrdValidator struct {
+	Log     logr.Logger
+	Client  client.Client
+	decoder *admission.Decoder
+}
+
+// Handle takes an incoming CRD create/update request and confirms that it is
+// a safe upgrade based on the crd.Validate() function call
+func (v *CrdValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := v.Log.V(1).WithName("crdhandler") // Default to non-verbose logs
+	incomingCrd := &apiextensionsv1.CustomResourceDefinition{}
+
+	err := v.decoder.Decode(req, incomingCrd)
+	if err != nil {
+		message := fmt.Sprintf("failed to decode CRD %q", req.Name)
+		log.V(0).Error(err, message)
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("%s: %w", message, err))
+	}
+
+	err = crd.Validate(ctx, v.Client, incomingCrd)
+	if err != nil {
+		message := fmt.Sprintf("failed to validate safety of %s for CRD %q: %s", req.Operation, req.Name, err)
+		log.V(0).Info(message)
+		return admission.Denied(message)
+	}
+
+	log.Info("admission allowed for %s of CRD %q", req.Name, req.Operation)
+	return admission.Allowed("")
+}
+
+// InjectDecoder injects a decoder for the CrdValidator.
+func (v *CrdValidator) InjectDecoder(d *admission.Decoder) error {
+	v.decoder = d
+	return nil
+}
```

Comment:
How do we feel about an options pattern implementation, .e.g `v.WithDecoder(d)` here?

Reply:
So the `InjectDecoder` actually implements an admission.DecoderInjector so that the decoder gets automatically injected. I got it from the `controller-runtime` example found [here](https://github.com/kubernetes-sigs/controller-runtime/blob/188b48630db275131d0fef1818933d8c2da96753/examples/builtins/validatingwebhook.go#L58).

Reply:
Alright if there's already an existing pattern in the community, then sticking with that is fine with me.